src/shm/pool/slot.rs

```diff
@@ -0,0 +1,412 @@
+//! A pool implementation based on buffer slots
+
+use std::io;
+use std::sync::{
+    atomic::{AtomicU8, AtomicUsize, Ordering},
+    Arc, Mutex, Weak,
+};
+
+use wayland_client::{
+    backend::InvalidId,
+    protocol::{wl_buffer, wl_shm, wl_surface},
+    Connection, Proxy,
+};
+
+use super::raw::RawPool;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CreateBufferError {
+    /// Protocol error.
+    #[error(transparent)]
+    Protocol(#[from] InvalidId),
+
+    /// Slot creation error.
+    #[error(transparent)]
+    Io(#[from] io::Error),
+
+    /// Pool mismatch.
+    #[error("Incorrect pool for slot")]
+    PoolMismatch,
+
+    /// Slot size mismatch
+    #[error("Buffer too large for slot")]
+    SlotTooSmall,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ActivateSlotError {
+    /// Buffer was already active
+    #[error("Buffer was already active")]
+    AlreadyActive,
+}
+
+#[derive(Debug)]
+pub struct SlotPool {
+    pub(crate) inner: RawPool,
```

Comment:
The shm pool is leaked as RawPool does not cleanup on drop.

In particular do note that a buffer **can** outlive it's pool protocol object

Reply:
I think that cleanup should be done in `RawPool`, unless there's a reason not to?

Reply:
correction: RawPool already does cleanup on drop.  Buffers continue to exist (and are usable, although you cannot access their canvas anymore) until they are dropped and released.